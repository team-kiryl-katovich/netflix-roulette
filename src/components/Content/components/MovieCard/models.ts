export type MovieCardProps = {
  id: number;
  title: string;
  gengre: string;
  year: number;
  imageUrl: string;
};
