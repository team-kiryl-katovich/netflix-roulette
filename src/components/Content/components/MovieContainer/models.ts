import { WithMoviesProps } from 'src/enhancers/withMovies';

export type MovieContainerComponentProps = WithMoviesProps;
