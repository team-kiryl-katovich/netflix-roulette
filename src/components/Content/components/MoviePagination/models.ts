import { Movie, Offset } from '@common/api';

export type MoviePaginationProps = {
  moviesCount: number;
  offset: Offset;
  pageSize: number;
  loading: boolean;
  children?: React.ReactNode;
  fetch: (offset: Offset) => void;
};

export type MoviePaginationState = {};
