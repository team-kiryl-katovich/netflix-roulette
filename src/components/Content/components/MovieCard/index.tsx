import * as React from 'react';
import { Link } from 'react-router-dom';

import { ScaleImage, Label } from '@common/components';
import { ROUTES } from '@common/routes';

import { MovieCardProps } from './models';

import './styles';

export const MovieCard = ({
  id, title, gengre, year, imageUrl,
}: MovieCardProps) => (
  <div className="movie-card">
    <Link to={`${ROUTES.MOVIE}/${id}`}>
      <ScaleImage src={imageUrl} ratio={140} />
      <div className="movie-card__description">
        <div className="movie-card-title">
          <Label className="movie-card-title__title" text={title} />
          <Label className="movie-card-title__gengre" text={gengre} />
        </div>
        <div className="movie-card-year">
          <Label className="movie-card-year__label" text={year.toString()} />
        </div>
      </div>
    </Link>
  </div>
);
