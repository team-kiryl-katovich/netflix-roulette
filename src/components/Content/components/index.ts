export * from './MovieCard';
export * from './MovieContainer';
export * from './MoviePagination';
