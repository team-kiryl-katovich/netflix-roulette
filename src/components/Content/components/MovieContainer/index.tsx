import * as React from 'react';

import { Loader } from '@common/components';
import { withMovies } from '@enhancers/withMovies';

import { MovieCard, MoviePagination } from '../';

import { MovieContainerComponentProps } from './models';

import './styles';

const pageSize = 9;

export const MovieContainerComponent = ({
  moviesData: { movies, loading, offset },
  searchMovies,
  moreMovies,
}: MovieContainerComponentProps) => {
  React.useEffect(() => {
    searchMovies({ limit: pageSize });
  }, []);
  if (loading && !movies.length) {
    return <Loader loading={loading} />;
  }
  return (
    <div className="movie-container">
      <MoviePagination
        moviesCount={movies.length}
        loading={loading}
        fetch={moreMovies}
        pageSize={pageSize}
        offset={offset}
      >
        {movies.map((movie) => (
          <MovieCard
            id={movie.id}
            title={movie.title}
            gengre={movie.genres.join(',')}
            year={new Date(movie.release_date).getFullYear()}
            imageUrl={movie.poster_path}
            key={movie.id}
          />
        ))}
      </MoviePagination>
    </div>
  );
};

export const MovieContainer = withMovies(MovieContainerComponent);
