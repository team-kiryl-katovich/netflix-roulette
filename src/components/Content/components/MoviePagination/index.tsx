import * as React from 'react';

import { MoviePaginationProps, MoviePaginationState } from './models';
import { Loader } from '@common/components';

export class MoviePagination extends React.Component<MoviePaginationProps, MoviePaginationState> {
  constructor(props: MoviePaginationProps) {
    super(props);
  }

  componentDidMount() {    
    window.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {    
    window.removeEventListener('scroll', this.handleScroll);
  }

  render() {
    const { loading, children } = this.props;
    return (
      <>
        {children}
        <Loader loading={loading} />
      </>
    );
  }

  private handleScroll = () => {
    if (!this.checkBottom()) {
      return;
    }
    if (this.props.loading) {
      return;
    }
    this.getMoreItems();
  };

  private checkBottom = (): boolean => {
    const { body, documentElement: html } = document;
    const docHeight = Math.max(
      body.scrollHeight,
      body.offsetHeight,
      html.clientHeight,
      html.scrollHeight,
      html.offsetHeight
    );
    const windowBottom = window.innerHeight + window.pageYOffset + 1;
    return windowBottom >= docHeight;
  };

  private getMoreItems() {
    const { moviesCount, offset, pageSize, fetch } = this.props;
    if (moviesCount < offset.limit) {
      fetch({
        limit: pageSize,
        offset: offset.offset + pageSize,
      });
    }
  }
}
