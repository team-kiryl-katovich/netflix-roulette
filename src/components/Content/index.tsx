import * as React from 'react';

import { ScaleImage } from '../../common/components';

import './styles';

// TODO: split into movie card and conteiner component with neccessary props

export const Content = () => (
  <div className="content">
    <div className="movie-container">
      {[1, 2, 3, 4, 5].map(index => (
        <div className="movie-card">
          <ScaleImage src={require(`@assets/data/movie-${index}.jpg`)} ratio={140} />
          <div className="movie-card__description">
            <div className="movie-card-title">
              <div className="movie-card-title__title">Pupl Fiction</div>
              <div className="movie-card-title__gengre">{`Action & Adventure`}</div>
            </div>
            <div className="movie-card-year">
              <div className="movie-card-year__label">2004</div>
            </div>
          </div>
        </div>
      ))}
    </div>
  </div>
);
