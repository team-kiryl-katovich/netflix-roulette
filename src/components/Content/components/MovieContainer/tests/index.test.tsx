import * as React from 'react';
import * as enzyme from 'enzyme';

import { Loader } from '@common/components';

import { MovieContainerComponentProps } from '../models';
import { MovieContainerComponent } from '..';

describe('[COMPONENTS] [CONTENT] [COMPONENTS] [MOVIE_CONTAINER_COMPONENT]', () => {
  const fakeProps: MovieContainerComponentProps = {
    moreMovies: jest.fn(),
    moviesData: {
      error: null,
      loading: false,
      movies: [
        {
          id: 1,
          genres: ['fakeGenre'],
          budget: 1,
          overview: 'fakeOverview',
          poster_path: 'fakePosterPath',
          release_date: '2019-01-01',
          revenue: 1,
          runtime: 1,
          tagline: 'fakeTagline',
          title: 'fakeTitle',
          vote_average: 1,
          vote_count: 1,
        },
      ],
      offset: {
        limit: 1,
        offset: 1,
      },
    },
    searchMovies: jest.fn(),
  };

  test('should render component', () => {
    const component = enzyme.shallow(<MovieContainerComponent {...fakeProps} />);
    expect(component).toMatchSnapshot();
  });

  test('should render loading component', () => {
    const props = {
      ...fakeProps,
      moviesData: {
        ...fakeProps.moviesData,
        loading: true,
        movies: [] as any,
      },
    };

    (props.searchMovies as jest.Mock).mockReset();

    const component = enzyme.mount(<MovieContainerComponent {...props} />);
    const loader = component.find(Loader);
    
    expect(loader.exists()).toBeTruthy();
    expect(props.searchMovies).toHaveBeenCalledTimes(1);
  });
});
