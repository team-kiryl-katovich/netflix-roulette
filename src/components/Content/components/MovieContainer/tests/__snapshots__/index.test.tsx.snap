// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENTS] [CONTENT] [COMPONENTS] [MOVIE_CONTAINER_COMPONENT] should render component 1`] = `
<div
  className="movie-container"
>
  <MoviePagination
    fetch={[MockFunction]}
    loading={false}
    moviesCount={1}
    offset={
      Object {
        "limit": 1,
        "offset": 1,
      }
    }
    pageSize={9}
  >
    <Component
      gengre="fakeGenre"
      id={1}
      imageUrl="fakePosterPath"
      key="1"
      title="fakeTitle"
      year={2019}
    />
  </MoviePagination>
</div>
`;
